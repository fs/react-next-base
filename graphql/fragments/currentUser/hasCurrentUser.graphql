fragment HasCurrentUser on CurrentUser {
  id
}
