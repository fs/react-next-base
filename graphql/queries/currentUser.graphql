#import "../fragments/userInfo.graphql"

query CurrentUser {
  me {
    ...UserInfo
  }
}
