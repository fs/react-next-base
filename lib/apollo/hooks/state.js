import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';

export const useCurrentUser = (prefetch = true) => {
  const { loading, data, error } = useQuery(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  const user = data?.me;

  return { user, loading, error };
};
