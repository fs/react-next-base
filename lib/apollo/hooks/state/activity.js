/* eslint-disable import/prefer-default-export */
import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

const getFormattedActivity = (data) => {
  return data.activities.edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event).color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({ beforeCursor, afterCursor, filterValue, pageSize = activityPageSizes[0] }) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data ? getFormattedActivity(data) : null,
    pageInfo: data ? data.activities.pageInfo : null,
    loading,
    error,
  };
};
