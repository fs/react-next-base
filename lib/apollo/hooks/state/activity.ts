import { useQuery } from '@apollo/client';
import Activities from 'graphql/queries/pages/activities.graphql';

import activityEvents from 'config/activityEvents';
import activityPageSizes from 'config/activityPageSizes';

import User from 'domain/User';
import { PageInfo, Activity } from 'interfaces/activityType';

type ActivityParams = {
  beforeCursor?: string;
  afterCursor?: string;
  filterValue?: string;
  pageSize: number;
};

type Edge = {
  cursor: string;
  node: {
    body: string;
    createdAt: string | Date;
    event: string;
    id: string | number;
    title: string;
    user: User;
  };
};

type ActivityQueryState = {
  activities: {
    edges: Edge[];
    pageInfo: PageInfo;
  };
};

type ActivityQueryVars = {
  events: string[];
  last?: number;
  before?: string;
  first?: number;
  after?: string;
};

const getFormattedActivity = (edges: Edge[]): Activity[] => {
  return edges.map(
    ({
      node: {
        id,
        title,
        body,
        createdAt,
        event,
        user: { firstName, lastName, email, avatarUrl },
      },
    }) => ({
      id,
      title,
      description: body,
      date: new Date(createdAt).toLocaleString(),
      color: activityEvents.find((e) => e.value === event)?.color,
      name: `${firstName} ${lastName}`,
      email,
      avatarUrl,
    }),
  );
};

export const useActivity = ({
  beforeCursor,
  afterCursor,
  filterValue,
  pageSize = activityPageSizes[0],
}: ActivityParams) => {
  const events = filterValue ? [filterValue] : activityEvents.map(({ value }) => value);
  const first = afterCursor || (!afterCursor && !beforeCursor) ? pageSize : undefined;
  const last = beforeCursor ? pageSize : undefined;

  const { loading, error, data } = useQuery<ActivityQueryState, ActivityQueryVars>(Activities, {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
    variables: {
      events,
      last,
      before: beforeCursor,
      first,
      after: afterCursor,
    },
  });

  return {
    activities: data?.activities?.edges ? getFormattedActivity(data?.activities.edges) : null,
    pageInfo: data?.activities?.pageInfo || null,
    loading,
    error,
  };
};
