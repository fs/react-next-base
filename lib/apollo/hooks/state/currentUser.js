import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';

export const useCurrentUser = (prefetch = true) => {
  const { data, loading, error } = useQuery(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  return {
    user: data?.me,
    loading,
    error,
  };
};
