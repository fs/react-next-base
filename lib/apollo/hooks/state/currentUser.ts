import { useQuery } from '@apollo/client';
import CurrentUser from 'graphql/queries/currentUser.graphql';
import User from 'domain/User';

type UserData = {
  me: User;
};

export const useCurrentUser = (prefetch = true) => {
  const { data, loading, error } = useQuery<UserData>(CurrentUser, {
    fetchPolicy: prefetch ? 'cache-first' : 'cache-only',
  });

  return {
    user: data?.me,
    loading,
    error,
  };
};
