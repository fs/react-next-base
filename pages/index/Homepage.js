import React from 'react';
import styled from 'styled-components';
import DefaultTemplate from 'components/templates/DefaultTemplate';

const Title = styled.h1`
  font-size: 2rem;
  text-align: center;
`;

const Home = () => (
  <DefaultTemplate>
    <Title>Welcome to React Next Base</Title>
  </DefaultTemplate>
);

export default Home;
