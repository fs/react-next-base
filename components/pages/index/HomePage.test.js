import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import renderWithApolloClient from '__tests__/helpers/renderWithApolloClient';
import renderWithTheme from '__tests__/helpers/renderWithTheme';

import { useCurrentUser } from 'lib/apollo/hooks/state/currentUser';

import Homepage from 'pages/index';

jest.mock('lib/apollo/hooks/state/currentUser');

describe('HomePage', () => {
  test('should render correctly', async () => {
    // Arrange
    const mockUseCurrentUser = jest.fn(() => ({
      loading: undefined,
      error: undefined,
      user: { id: '1', email: 'user@mail.ru' },
    }));
    useCurrentUser.mockImplementation(mockUseCurrentUser);

    // Act
    render(renderWithTheme(renderWithApolloClient(<Homepage />)));

    // Assert
    expect(mockUseCurrentUser).toHaveBeenCalled();
  });
});
