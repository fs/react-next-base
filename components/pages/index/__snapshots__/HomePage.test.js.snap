// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render correctly 1`] = `
.c0 {
  font-size: 2rem;
  text-align: center;
}

@media (max-width:576px) {

}

@media (max-width:1200px) {

}

<div
  class="styled__PageContent-sc-1bxgj0r-1"
  data-testid="page-content"
>
  <h1
    class="c0 capitalize"
    data-cy="welcome-page"
  >
    Welcome to React Next Base
  </h1>
</div>
`;
